### ISSUE TITLE
Feature Request: allow override of dockerd options 

### ISSUE THREAD

**Issue by robbyoconnor** [geerlingguy/ansible-role-docker#90]
I love your Ansible Roles -- but the major no-go for me is the fact I can't override the storage driver easily.

**Comment by llbbl** [comment]
```
# defaults
docker_default_config:
  storage-driver: overlay2
  log-level: info

# playbook
- name: apply default daemon config
  set_fact:
    docker_json: "{{ docker_default_config | combine(docker_config | default({})) }}"

- name: ensure daemon config file is present
  copy:
    content: |
      {{ docker_json | to_nice_json(indent=2) }}
    dest: /etc/docker/daemon.json
  notify: restart docker

```

how mongrelion is doing it.

**Comment by llbbl** [comment]
FYI devicemapper is depreciated and causing issues. 

https://github.com/docker/cli/pull/1424

**Comment by stegr04** [comment]
This role is installing v18.09.5 for me as of this post. I'm not sure what might have changed since this issue was originally opened but I figured I would share my findings. 

I was looking for the same thing (change storage driver) because docker info was showing it was using devicemapper after installing docker with this role. I found this issue and the solutions discussed/created. Awesome!

I started testing Rickkwa's solution (thank you!). It was working as advertised. Then I started testing without the storage options settings (after manually removing docker). I noticed it was still using overlay2. I wasn't expecting that. I then realized the uninstall wasn't removing the /etc/docker/daemon.json file. 

Good time for some backgroun... all of the hosts I am testing with had docker previously installed. I think the system was setup this way because prior versions of docker required Kernel v4 on CentOs. 

To make sure I wasn't picking up on anything from prior setups I removed /etc/docker and /var/lib/docker. I tested with Rickkwa's again (no storage options again) and it still showed overlay2, remove everything again, tested geerlingguy's role and found it using overlay2. 

To see if there was something else I might have been missing I tested geerlingguy's role on a server I haven't touched except to uninstall all docker packages on the machine. It did not have any /etc/docker/daemon.json file but it did still have /var/lib/docker. When I tested against this node docker info showed me it was using devicemapper. So I did the manual uninstall again, cleaned up the /var/lib/docker and /etc/docker/key.json, ran geerlingguy's role and it's using overlay2.

**Comment by robbyoconnor** [comment]
@geerlingguy

**Comment by robbyoconnor** [comment]
This issue is not resolved

**Comment by garymoon** [comment]
We're after this also, and will be copying tasks from one of the PRs to get around it for now. Would be happy to contribute fixes/cleanup to one of the PRs to get it merged.

**Comment by onezeroecho** [comment]
Same here, when using Molecule for testing an Ansible playbook which runs Docker containers you have to use `storage-driver: vfs` because apparently Docker does not support `aufs` over `aufs`.

**Comment by robbyoconnor** [comment]
> Same here, when using Molecule for testing an Ansible playbook which runs Docker containers you have to use `storage-driver: vfs` because apparently Docker does not support `aufs` over `aufs`.

@onezeroecho why not `overlay2`?

**Comment by onezeroecho** [comment]
@robbyoconnor Doesn't work for my use case, I get the following output during `molecule test` when I set the storage driver to `overlay2` instead of `vfs`:

```
TASK [Get Docker container information] ***************************************
failed: [instance] (item=app_rocketchat_1) => {"ansible_loop_var": "item", "changed": false, "item":
"app_rocketchat_1", "msg": "Error connecting: Error while fetching server API version:
('Connection aborted.', ConnectionRefusedError(111, 'Connection refused'))"}
```

See also [Continuous Infrastructure with Ansible, Molecule & TravisCI](https://blog.codecentric.de/en/2018/12/continuous-infrastructure-ansible-molecule-travisci/) where they had to do the same thing.

**Comment by robbyoconnor** [comment]
@onezeroecho interesting

**Comment by robbyoconnor** [comment]
bump

**Comment by robbyoconnor** [comment]
bump

**Comment by damianoneill** [comment]
Hi @geerlingguy can you provide a comment please on whether this feature is something that should be added to this role?  It's unclear from looking at the thread history (its been running for two years) whether there is any appetite to update the role to support this.

**Comment by geerlingguy** [comment]
@damianoneill  I do think this kind of functionality would be nice to have in this role.

**Comment by geerlingguy** [comment]
Left one review on the associated PR, other than that this will be good to go!

**Comment by robbyoconnor** [comment]
bump

**Comment by geerlingguy** [comment]
There's also this PR: https://github.com/geerlingguy/ansible-role-docker/pull/261

**Comment by robbyoconnor** [comment]
@geerlingguy just keeping PRs that I have a stake in from closing :)

**Comment by jonathantullett** [comment]
Is #261 ready for merging @geerlingguy? It would be great to get this merged.

**Comment by geerlingguy** [comment]
This landed in the 4.0.0 release via #256—thanks to everyone for your patience in getting this merged!
